### PR TITLE
chore: support iam role prefix and boundary

### PR DESCRIPTION
**Pull request by tyyzqmf**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [x] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts [hunk 1]
```diff
@@ -19043,6 +19043,9 @@ describe('SQL Builder test', () => {
           union all
           select
             (CURRENT_DATE - INTERVAL '3 day')::date as event_date
+          union all
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts:19046]: Why are you changing this?

File: src/control-plane/backend/lambda/listen-stack-status/listen-tools.ts [hunk 2]
```diff
@@ -37,6 +38,7 @@ const docClient = DynamoDBDocumentClient.from(ddbClient, {
 
 const clickStreamTableName = process.env.CLICKSTREAM_TABLE_NAME ?? '';
 const prefixTimeGSIName = process.env.PREFIX_TIME_GSI_NAME ?? '';
+export const stackPrefix = process.env.IAM_ROLE_PREFIX ? `${process.env.IAM_ROLE_PREFIX}-${SolutionInfo.SOLUTION_SHORT_NAME}` : SolutionInfo.SOLUTION_SHORT_NAME;
```
Comment (zxkane) [inline src/control-plane/backend/lambda/listen-stack-status/listen-tools.ts:41]: Why not use the method `getStackPrefix()`?

File: src/control-plane/backend/stack-action-state-machine-construct.ts [hunk 3]
```diff
@@ -47,6 +49,8 @@ export class StackActionStateMachine extends Construct {
   constructor(scope: Construct, id: string, props: StackActionStateMachineProps) {
     super(scope, id);
 
+    const stackNamePrefix = props.iamRolePrefix ? `${props.iamRolePrefix}-${SolutionInfo.SOLUTION_SHORT_NAME}` : SolutionInfo.SOLUTION_SHORT_NAME;
```
Comment (zxkane) [inline src/control-plane/backend/stack-action-state-machine-construct.ts:52]: Why not use the method `getStackPrefix()`?

File: src/main.ts [hunk 4]
```diff
@@ -41,6 +43,9 @@ function stackSuppressions(stacks: Stack[], suppressions: NagPackSuppression[])
   });
 }
 
+const iamRolePrefix = app.node.tryGetContext('rolePrefix');
```
Comment (zxkane) [inline src/main.ts:46]: move to use CfnParameter instead of CDK context, which only works on deploying from source code

File: src/main.ts [hunk 5]
```diff
@@ -299,6 +313,84 @@ class CNLambdaFunctionAspect implements IAspect {
...
+  }
+
+  public visit(node: IConstruct): void {
+    if (CfnResource.isCfnResource(node) && node.cfnResourceType === 'AWS::IAM::Role') {
+      node.addPropertyOverride('PermissionsBoundary', `arn:${Aws.PARTITION}:iam::${Aws.ACCOUNT_ID}:policy/${this.permissionBoundary}`);
+    }
+  }
+}
+if (app.node.tryGetContext('rolePermissionBoundary')) {
+  Aspects.of(app).add(new RolePermissionBoundaryAspect(app.node.tryGetContext('rolePermissionBoundary')));
```
Comment (zxkane) [inline src/main.ts:391]: move to use CfnParameter instead of CDK context, which only works on deploying from source code

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts (hunk 1 above)
Comment (tyyzqmf) [reply src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts:19046]: fixed.

File: src/main.ts (hunk 4 above)
Comment (tyyzqmf) [reply src/main.ts]: done.

File: src/service-catalog-appregistry-stack.ts [hunk 6]
```diff
@@ -75,5 +76,11 @@ export class ServiceCatalogAppregistryStack extends Stack {
       value: application.applicationArn,
       condition: serviceAvailableRegion,
     });
+
+    // Add IAM role permission boundary aspect
+    const {
+      iamRoleBoundaryArnParam,
+    } = Parameters.createIAMRolePrefixAndBoundaryParameters(this);
+    Aspects.of(this).add(new RolePermissionBoundaryAspect(iamRoleBoundaryArnParam.valueAsString));
```
Comment (zxkane) [inline src/service-catalog-appregistry-stack.ts:84]: Currently, it's useless. We don't create an IAM role in this stack.

**Review by zxkane — COMMENTED** [review]
Do you pass the role bounary to the stacks of the data pipeline if the bounary is given for the web console stack?

If so, please add a test case for it.

**Review by zxkane — CHANGES_REQUESTED** [review]
Do you pass the role bounary to the stacks of the data pipeline if the bounary is given for the web console stack?

If so, please add a test case for it.

**Comment by tyyzqmf** [comment]
> Do you pass the role bounary to the stacks of the data pipeline if the bounary is given for the web console stack?
> 
> If so, please add a test case for it.

The parameter has been added to pipeline stack and the corresponding test case.

**Review by zxkane — APPROVED** [review]

**Pull request merged**